staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,143 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscoveries.
+// The shape is chosen to be very similar to "normal" api endpoints and work well with our type system
+// This is what is returned from the /discovery/v2 endpoint.
+type APIGroupDiscoveryList struct {
+	TypeMeta `json:",inline"`
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Groups is the list of groups for discovery.
+	// +listType=map
+	// +listMapKey=name
+	Groups []APIGroupDiscovery `json:"groups" protobuf:"bytes,2,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscovery holds information about which resources are being served for all version of the API Group.
+type APIGroupDiscovery struct {
+	TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// The only field completed will be Name.  For instance, resourceVersion will be empty.
+	// Name is the name of the API group whose discovery information is presented here.
+	// Name is allowed to be "" to represent the legacy, ungroupified resources.
+	// This is included to make the standard deserialization easier.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Versions are the versions supported in this group.
+	// +listType=map
+	// +listMapKey=version
+	Versions []APIVersionDiscovery `json:"versions,omitempty" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// APIVersionDiscovery holds a list of APIResourceDiscovery types that are served for a particular version within an API Group.
+type APIVersionDiscovery struct {
+	// Version is the name of the version within a group version.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// +listType=map
+	// +listMapKey=resource
+	Resources []APIResourceDiscovery `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// Most recently observed status of the APIVersionDiscovery.
+	// This data may not be up to date.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status APIVersionDiscoveryStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
This bit moved.  The reason I had this directly on `APIGroupDiscovery` was so that "normal" handlers that inspect resources for conditions would be able to find them.  I see the value of programatically knowing that a particular version's data is stale.  Perhaps we could also maintain a list of versions in a regular .Status?

Reply:
> so that "normal" handlers that inspect resources for conditions would be able to find them

Can you talk a bit in more detail about this use case?  We moved this because `APIService` is defined at the GroupVersion level; so it did not make as much sense to us to have a single `Stale` (or other) condition for the entire Group.

Do I understand your suggestion correctly that we should instead have a list of conditions for each version in the Group-level Status?

Reply:
IIRC, an APIService is partitioned by group-version. We have no guarantee that every version in a group is served by the same APIService right? Wouldn't it make more sense for this to be nested within a version object instead of the group?

> The reason I had this directly on APIGroupDiscovery was so that "normal" handlers that inspect resources for conditions would be able to find them

Can you clarify this comment?

Reply:
> Do I understand your suggestion correctly that we should instead have a list of conditions for each version in the Group-level Status?

I read it as we should have both the stale information for each GroupVersion along with a list of the versions that are stale at the Group level. I'm generally concerned about an API that would repeat the same information twice.

Reply:
Having human status that would interact with "normal" tools and machine information for consumption of this specific API seems worthwhile to me.  We do this often, for instance the deployment condition "Available" is calculated based on available, replicas, and max-unavailable.  All of which are observable from the instance for machine consumability, while the condition is written for human consumption.

Reply:
We're trying to reach consensus on whether to duplicate the information we publish to the aggregated discovery group for a machine-readable/human-readable split. cc @liggitt for an additional opinion.

This is for marking apiservices as stale if they are unreachable within a certain period of time. We have `staleVersions` as a list of versions within a group that are stale, and also `conditions` emitting a human-readable message of the versions that are stale. The information is kind of duplicated between the two fields which is not ideal, but also in a way targeting different consumers of the API. 

The API looks something like this for two stale versions within a group of three versions: 

```
type: APIGroupDiscovery
name: groupname
versions:
  - v1
  - v2
  - v3
status:
  conditions:
    - type: stale
      message: failed to read group/v1 discovery <url here> due to error: %v.
               failed to read group/v2 discovery <url here> due to error: %v.
  staleVersions:
    - v1
    - v2
```

Reply:
OK I think I had misread that. I think having both the version AND the error message is perfectly fine.
I don't know if I really love the "stale" condition for the entire group when it actually only applies to specific versions, I think that's a little misleading. Why don't we have the condition per version? Then we don't even need the duplication. The problem with that is that objects don't look like native k8s objects then, e.g.
```
type: APIGroupDiscovery
name: groupname
versions:
  - v1
  - v2
  - v3
status:
  conditions:
    - type: stale
      message: failed to read group/v1 discovery <url here> due to error: %v.
               failed to read group/v2 discovery <url here> due to error: %v.

Reply:
it seems like there's a couple motivations here:
1. put something at `status.conditions` so generic handlers can observe when the stale condition is present/absent
   * this seems maybe useful for duck-typing/generic condition handlers (though if this is not a watchable API, some of those generic controllers will not be happy about waiting for the condition)
   * the whole object type is sort of reflecting observed state, so splitting `versions` and `status` as top-level peer fields is weirdly misleading
2. return stale/error information with coherent granularity
  * matching the granularity we're aggregating discovery at (version-level granularity) makes sense to me
  * maintaining parallel `versions...` and `status.staleVersions` lists callers have to correlate themselves seems weird for an API that is effectively all status
  * if there is specific information we want for each version (is it stale? when did it go stale? what was the last error?), I don't know that I'd embed generic []metav1.Condition lists for each version, especially if this isn't a writeable API that arbitrary writers can collaborate on

Reply:
I would have balanced the choice differently, but I'll live with embedding staleness on the version.